### ISSUE TITLE
Random Exception between Integrity Error and FlushError

### ISSUE THREAD

**Issue by achavhantexada** [sqlalchemy/sqlalchemy#4662]
Unittest an API, where test code fixture is trying to create a resource and then get the response to verify the previous step, and then again to try to create the same resource again, SqlAlchemy is randomizing the exception between Integrity and Flush Error. 

Libs: 
Flask==1.0.2
Flask-Cors==3.0.6
Flask-SQLAlchemy==2.3.2
SQLAlchemy==1.2.7
Werkzeug==0.14.1

I have been working on that project for the last 10 month, all the time I have seen only Integrity error, not flash error. 

            new_item.before_created(request_data)
            db.session.add(new_item)
            new_item.after_created(request_data)
            db.session.commit()

            headers = {'Location': new_item.get_url()} if new_item.get_url() else None
            return (jsonify({}), 201, headers) if headers else (jsonify({}), 201)
        except KeyError as e:
            raise ValidationError(
                ErrorCode.MISSING_MANDATORY_DATA,
                error_msg=f"the {e.args[0]} field is required.",
            )
        except ValueError as e:
            raise ValidationError(ErrorCode.INVALID_REQUEST_DATA, error_msg=e.args[0])
        except BadRequest:
            raise ValidationError(ErrorCode.INVALID_REQUEST_FORMAT)
        except IntegrityError:
            raise ValidationError(ErrorCode.INTEGRITY_ERROR, status_code=409)
        except Conflict as e:
            raise ValidationError(e.description, status_code=409)
        finally:
            db.session.rollback()

TEST CODE:

        def test_create_new_product(self):
              Product.delete(self.product_no, True)

        # create a asset with the request data.
        response = self.client.post(
            "/assets/",
            content_type="application/json",
            data=json.dumps(self.default_req_data),
        )
        res_data = json.loads(response.data.decode())
        self.assertEqual({}, res_data)
        self.assertEqual(response.status_code, 201)

        # check the asset that we already created.
        response = self.client.get(
            "/assets/{0}".format(self.product_no), content_type="application/json"
        )
        res_data = json.loads(response.data.decode())
        self.assertEqual(response.status_code, 200)

        for key in self.default_req_data.keys():
            self.assertEqual(self.default_req_data[key], res_data[key])

        # try create duplicated asset
        response = self.client.post(
            "/assets/",
            content_type="application/json",
            data=json.dumps(self.default_req_data),
        )
        res_data = json.loads(response.data.decode())
        self.assertEqual(response.status_code, 409)

This fixture is called by several other fixtures but all them get Integrity error except one of them is getting flushError.  

If I run all the test together I am facing this issue, alone it works fine.

**Comment by zzzeek** [comment]
Hi  there -

I can't work without a self-contained test case that runs and shows the stack trace.   Also, need the stack trace.  thanks.

**Comment by achavhantexada** [comment]
> 2019-05-03 17:57:43,969 ERROR: ['Traceback (most recent call last):\n', '  File "/var/jenkins_home/workspace/_delete-receving-parts-test-X5XT5MYZ6RD3JLIMHBSOPHMLGJX7LFGQ3H4H5WCKFOAULSQRDNLQ/env/lib/python3.6/site-packages/flask/app.py", line 1813, in full_dispatch_request\n    rv = self.dispatch_request()\n', '  File "/var/jenkins_home/workspace/_delete-receving-parts-test-X5XT5MYZ6RD3JLIMHBSOPHMLGJX7LFGQ3H4H5WCKFOAULSQRDNLQ/env/lib/python3.6/site-packages/flask/app.py", line 1799, in dispatch_request\n    return self.view_functions[rule.endpoint](**req.view_args)\n', '  File "/var/jenkins_home/workspace/_delete-receving-parts-test-X5XT5MYZ6RD3JLIMHBSOPHMLGJX7LFGQ3H4H5WCKFOAULSQRDNLQ/env/lib/python3.6/site-packages/flask/views.py", line 88, in view\n    return self.dispatch_request(*args, **kwargs)\n', '  File "/var/jenkins_home/workspace/_delete-receving-parts-test-X5XT5MYZ6RD3JLIMHBSOPHMLGJX7LFGQ3H4H5WCKFOAULSQRDNLQ/env/lib/python3.6/site-packages/flask/views.py", line 158, in dispatch_request\n    return meth(*args, **kwargs)\n', '  File "/var/jenkins_home/workspace/_delete-receving-parts-test-X5XT5MYZ6RD3JLIMHBSOPHMLGJX7LFGQ3H4H5WCKFOAULSQRDNLQ/srm_rest_api/decorators.py", line 246, in wrapped\n    new_item.after_created(request_data)\n', '  File "/var/jenkins_home/workspace/_delete-receving-parts-test-X5XT5MYZ6RD3JLIMHBSOPHMLGJX7LFGQ3H4H5WCKFOAULSQRDNLQ/srm_rest_api/asset/models.py", line 365, in after_created\n    self.create_subresource(request_data)\n', '  File "/var/jenkins_home/workspace/_delete-receving-parts-test-X5XT5MYZ6RD3JLIMHBSOPHMLGJX7LFGQ3H4H5WCKFOAULSQRDNLQ/srm_rest_api/asset/models.py", line 352, in create_subresource\n    Barcode.create(self.rspf_prod_n, request_data.get("barcodes"))\n', '  File "/var/jenkins_home/workspace/_delete-receving-parts-test-X5XT5MYZ6RD3JLIMHBSOPHMLGJX7LFGQ3H4H5WCKFOAULSQRDNLQ/srm_rest_api/asset/models.py", line 427, in create\n    if cls.query.filter_by(**filters).one_or_none():\n', '  File "/var/jenkins_home/workspace/_delete-receving-parts-test-X5XT5MYZ6RD3JLIMHBSOPHMLGJX7LFGQ3H4H5WCKFOAULSQRDNLQ/env/lib/python3.6/site-packages/sqlalchemy/orm/query.py", line 2854, in one_or_none\n    ret = list(self)\n', '  File "/var/jenkins_home/workspace/_delete-receving-parts-test-X5XT5MYZ6RD3JLIMHBSOPHMLGJX7LFGQ3H4H5WCKFOAULSQRDNLQ/env/lib/python3.6/site-packages/sqlalchemy/orm/query.py", line 2924, in __iter__\n    self.session._autoflush()\n', '  File "/var/jenkins_home/workspace/_delete-receving-parts-test-X5XT5MYZ6RD3JLIMHBSOPHMLGJX7LFGQ3H4H5WCKFOAULSQRDNLQ/env/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 1434, in _autoflush\n    self.flush()\n', '  File "/var/jenkins_home/workspace/_delete-receving-parts-test-X5XT5MYZ6RD3JLIMHBSOPHMLGJX7LFGQ3H4H5WCKFOAULSQRDNLQ/env/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 2254, in flush\n    self._flush(objects)\n', '  File "/var/jenkins_home/workspace/_delete-receving-parts-test-X5XT5MYZ6RD3JLIMHBSOPHMLGJX7LFGQ3H4H5WCKFOAULSQRDNLQ/env/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 2380, in _flush\n    transaction.rollback(_capture_exception=True)\n', '  File "/var/jenkins_home/workspace/_delete-receving-parts-test-X5XT5MYZ6RD3JLIMHBSOPHMLGJX7LFGQ3H4H5WCKFOAULSQRDNLQ/env/lib/python3.6/site-packages/sqlalchemy/util/langhelpers.py", line 66, in __exit__\n    compat.reraise(exc_type, exc_value, exc_tb)\n', '  File "/var/jenkins_home/workspace/_delete-receving-parts-test-X5XT5MYZ6RD3JLIMHBSOPHMLGJX7LFGQ3H4H5WCKFOAULSQRDNLQ/env/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 187, in reraise\n    raise value\n', '  File "/var/jenkins_home/workspace/_delete-receving-parts-test-X5XT5MYZ6RD3JLIMHBSOPHMLGJX7LFGQ3H4H5WCKFOAULSQRDNLQ/env/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 2344, in _flush\n    flush_context.execute()\n', '  File "/var/jenkins_home/workspace/_delete-receving-parts-test-X5XT5MYZ6RD3JLIMHBSOPHMLGJX7LFGQ3H4H5WCKFOAULSQRDNLQ/env/lib/python3.6/site-packages/sqlalchemy/orm/unitofwork.py", line 391, in execute\n    rec.execute(self)\n', '  File "/var/jenkins_home/workspace/_delete-receving-parts-test-X5XT5MYZ6RD3JLIMHBSOPHMLGJX7LFGQ3H4H5WCKFOAULSQRDNLQ/env/lib/python3.6/site-packages/sqlalchemy/orm/unitofwork.py", line 556, in execute\n    uow\n', '  File "/var/jenkins_home/workspace/_delete-receving-parts-test-X5XT5MYZ6RD3JLIMHBSOPHMLGJX7LFGQ3H4H5WCKFOAULSQRDNLQ/env/lib/python3.6/site-packages/sqlalchemy/orm/persistence.py", line 156, in save_obj\n    base_mapper, states, uowtransaction\n', '  File "/var/jenkins_home/workspace/_delete-receving-parts-test-X5XT5MYZ6RD3JLIMHBSOPHMLGJX7LFGQ3H4H5WCKFOAULSQRDNLQ/env/lib/python3.6/site-packages/sqlalchemy/orm/persistence.py", line 319, in _organize_states_for_save\n    state_str(existing)))\n', "sqlalchemy.orm.exc.FlushError: New instance <Product at 0x7fde864d34a8> with identity key (<class 'srm_rest_api.asset.models.Product'>, (1, '    SH_TEST1'), None) conflicts with persistent instance <Product at 0x7fde864d3898>\n"] [in /var/jenkins_home/workspace/_delete-receving-parts-test-X5XT5MYZ6RD3JLIMHBSOPHMLGJX7LFGQ3H4H5WCKFOAULSQRDNLQ/srm_rest_api/__init__.py:165]


This trace is when SqlAlchemy throws flusherror which is unusual because in the normal situation I get Integrity Error.

**Comment by zzzeek** [comment]
the error means you've loaded a Product with a particular primary key into the Session, then you made another Product object, gave it the same primary key, and then added it to the Session again as a pending object.   the ORM catches these since it is trying to see if you intended to delete the old object.

**Comment by achavhantexada** [comment]
But all the time Test case gets Integrity Error, not the FlushError. why SqlAchemy is throwing FlushError, because, this "test_create_new_product" fixture is used by other fixtures as well, and All the time Fixtures get Integrity errors, not flush errors except one this case. And it is random, not constant.

**Comment by zzzeek** [comment]
> But all the time Test case gets Integrity Error, not the FlushError. why SqlAchemy is throwing FlushError, because, 

because you have a Product() with the same primary key already loaded into the Session in these cases.


> this "test_create_new_product" fixture is used by other fixtures as well, and All the time Fixtures get Integrity errors, 

in the other cases, you do not have a Product() with the same primary key already loaded into the Session.

> not flush errors except one this case. And it is random, not constant.

This is because your test suite is randomly having a Product() with the same primary key, or not, as the new Product() you are trying to add to the Session.  Try making sure the Session is empty on each new test case.

**Comment by achavhantexada** [comment]
> This is because your test suite is randomly having a Product() with the same primary key, or not, as the new Product() you are trying to add to the Session. Try making sure the Session is empty on each new test case.

As I mentioned before I am using Flask and Flask-Sqlalchmey, and also for unittest, I am using Flask test client. I don't know it is my code fault or test client is keeping the old session alive. but if you see the code in finally section I am calling rollback as well. I believe, after the rollback call, new session transaction begins?

**Comment by zzzeek** [comment]
rollback is not enough to remove a Product that is already in there, it only expires the state of a Product instance.

Here is a demonstration of all the conditions that you describe.   I would recommend looking inside of session.identity_map to see what Product instances are already there and why they might be hanging around.

```
from sqlalchemy import *
from sqlalchemy.orm import *
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.ext.declarative import declared_attr

Base = declarative_base()


class Product(Base):
    __tablename__ = 'product'

    id = Column(Integer, primary_key=True)

e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

s.add(Product(id=1))
s.commit()
s.close()

# here's a test.

p1 = s.query(Product).filter_by(id=1).first()

s.rollback()

print(list(s.identity_map.values()))  # note the Product is still there

s.add(Product(id=1))
try:
    s.commit()  # <-- FlushError
except Exception as err:
    print(err)

# close the Session, will remove all objects
s.close()


print(list(s.identity_map.values()))  # now it's empty


s.add(Product(id=1))
try:
    s.commit()  # <-- IntegrityError
except Exception as err:
    print(err)


```

**Comment by zzzeek** [comment]
I know what it is you want to achieve here and I think the logic would be more consistent raising IntegrityError in all cases.

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **master** branch:

**Turn FlushError for identity already exists into a warning.** https://gerrit.sqlalchemy.org/1256

**Comment by achavhantexada** [comment]
I want to add something in this conversation that, I was not calling the close method, that's true because I thought Flask-Sqlachmey might be handling in internal but I saw the code and did not find any close call in their logic. Therefore, I added close methods during the creation and updation of resources, and it works fine now, I have no issue with the dangling objects in sessions. I believe I was following anti-pattern before.
